qiskit/scheduler/methods/basic.py

```diff
@@ -0,0 +1,154 @@
+    Returns:
+        A final schedule, pulses occuring as early as possible
+    """
+    circ_pulse_defs = translate_gates_to_pulse_defs(circuit, schedule_config)
+
+    def update_times(inst_qubits: List[int], time: int = 0) -> None:
+        """Update the time tracker for all inst_qubits to the given time."""
+        for q in inst_qubits:
+            qubit_time_available[q] = time
+
+    sched = Schedule()
+    qubit_time_available = defaultdict(int)
+
+    for circ_pulse_def in circ_pulse_defs:
+        time = max(qubit_time_available[q] for q in circ_pulse_def.qubits)
+        if isinstance(circ_pulse_def.schedule, Barrier):
+            update_times(circ_pulse_def.qubits, time)
+        else:
+            sched |= circ_pulse_def.schedule << time
+            update_times(circ_pulse_def.qubits, time + circ_pulse_def.schedule.duration)
+    return sched
+
+
+def as_late_as_possible(circuit: QuantumCircuit,
+                        schedule_config: ScheduleConfig) -> Schedule:
+    """
+    Return a pulse Schedule which nails down the timing between circuit element schedules by
+    scheduling pulses as late as possible on a resource, where resources are binned by qubit. The
+    timing between instructions within each circuit schedule are respected.
+
+    This method should improves the outcome fidelity over ASAP scheduling, because we may
+    maximize the time that the qubit remains in the ground state.
+
+    Args:
+        circuit: The quantum circuit to translate
+        schedule_config: Backend specific parameters used for building the Schedule
+    Returns:
+        A final schedule, pulses occuring as late as possible
+    """
+    circ_pulse_defs = translate_gates_to_pulse_defs(circuit, schedule_config)
+
+    def update_times(inst_qubits: List[int], shift: int = 0) -> None:
+        """Update the time tracker for all inst_qubits to the given time."""
+        for q in qubit_available_until.keys():
+            if q in inst_qubits:
+                qubit_available_until[q] = 0
+            else:
+                qubit_available_until[q] += shift
+
+    sched = Schedule()
+    qubit_available_until = defaultdict(lambda: float("inf"))
+
+    for circ_pulse_def in reversed(circ_pulse_defs):
+        if isinstance(circ_pulse_def.schedule, Barrier):
+            update_times(circ_pulse_def.qubits)
+        else:
+            cmd_sched = circ_pulse_def.schedule
+            cmd_start_time = (min([qubit_available_until[q] for q in circ_pulse_def.qubits])
+                              - cmd_sched.duration)
+            if cmd_start_time == float("inf"):
+                cmd_start_time = 0
+            shift_amount = max(0, -cmd_start_time)
+            cmd_start_time = max(cmd_start_time, 0)
+            sched = pulse.ops.union((shift_amount, sched), (cmd_start_time, cmd_sched))
+            update_times(circ_pulse_def.qubits, shift_amount)
+    return sched
+
+
+def translate_gates_to_pulse_defs(circuit: QuantumCircuit,
```

Comment:
We should be clear here that is can only schedule a circuit with a single set of measures. Ie., if you measure the same qubit twice the output schedule will have the undesired behaviour. Additionally, the results will be stored in the MemorysSlot of the same index as the qubit.

Reply:
What is the "undesired behavior" you're talking about? Why can't we schedule with multiple measures on a qubit?

Reply:
Good point about MemSlots

Reply:
How are memory slots specified otherwise? Does the cmd_def take Memory and Register slots for `.get('measure', ...)`?

Reply:
This is a failure of the `CmdDef` currently. Right now the acquire statement is hardcoded by the backend in a similar manner as the pulse channels. This is not as correct as `MemorySlot`s and `RegisterSlot`s are much more fungible than channels.  We should consider how we would remap channels/memoryslots/registerslots.

Reply:
I think this is wrapped up in #2704 now